src/nbl/video/COpenGL_Queue.h

```diff
@@ -27,6 +28,45 @@ class COpenGL_Queue final : public IGPUQueue
     private:
         static inline constexpr bool IsGLES = (FunctionTableType::EGL_API_TYPE == EGL_OPENGL_ES_API);
 
+        static inline GLbitfield pipelineStageFlagsToMemBarrierBits(asset::E_PIPELINE_STAGE_FLAGS flags)
+        {
+            constexpr GLbitfield VertexInputBits = GL_VERTEX_ATTRIB_ARRAY_BARRIER_BIT | GL_ELEMENT_ARRAY_BARRIER_BIT;
+            constexpr GLbitfield AnyShaderStageCommonBits = GL_UNIFORM_BARRIER_BIT | GL_TEXTURE_FETCH_BARRIER_BIT | GL_SHADER_IMAGE_ACCESS_BARRIER_BIT | GL_SHADER_STORAGE_BARRIER_BIT;
+            constexpr GLbitfield TransferBits = GL_PIXEL_BUFFER_BARRIER_BIT | GL_TEXTURE_UPDATE_BARRIER_BIT | GL_BUFFER_UPDATE_BARRIER_BIT;
+            constexpr GLbitfield HostBits = TransferBits | IOpenGL_FunctionTable::CLIENT_MAPPED_BUFFER_BARRIER_BIT;
+            constexpr GLbitfield AllGraphicsBits = GL_COMMAND_BARRIER_BIT | VertexInputBits | AnyShaderStageCommonBits;
+            constexpr GLbitfield AllCommandsBits = GL_ALL_BARRIER_BITS;
+            constexpr uint32_t PipelineStageCount = 14u;
+            const GLbitfield bits[PipelineStageCount] = {
```

Comment:
you need to make a different array dependeing on source stage or destination stage.

For Source:
```cpp
const GLbitfield producerBits[PipelineStageCount] = { // srcStageMask
GL_ALL_BARRIER_BITS,
GL_COMMAND_BARRIER_BIT,
GL_COMMAND_BARRIER_BIT|VertexInputBits, // vertex input stage is later than command stage
AllGraphicsBits, // vertex shader stage is later than vertex input stage
AllGraphicsBits, // control shader stage is later than vertex shader stage, but cannot consume any new resource types
AllGraphicsBits, // evaluation shader stage is later than control shader stage, but cannot consume any new resource types
AllGraphicsBits, // geometry shader stage is later than evaluation shader stage, but cannot consume any new resource types
AllGraphicsBits, // fragment shader stage is later than early fragment test stage, but cannot consume any new resource types
AllGraphicsBits, // early fragment test stage is later than geometry shader stage, but cannot consume any new resource types
AllGraphicsBits, // late fragment test stage is later than fragment shader stage, but cannot consume any new resource types
AllGraphicsBits, // color attachment output stage is later than late fragment test stage, but cannot consume any new resource types
GL_COMMAND_BARRIER_BIT|AnyShaderStageCommonBits, // compute shader stage is later than command stage
TRANSFER,
0 // bottom of pipe produces nothing
};
```

For Destination
```cpp
constexpr GLbitfield VertexInputBits = GL_VERTEX_ATTRIB_ARRAY_BARRIER_BIT | GL_ELEMENT_ARRAY_BARRIER_BIT;
constexpr GLbitfield AnyShaderStageCommonBits = GL_UNIFORM_BARRIER_BIT | GL_TEXTURE_FETCH_BARRIER_BIT | GL_SHADER_IMAGE_ACCESS_BARRIER_BIT | GL_SHADER_STORAGE_BARRIER_BIT;
constexpr GLbitfield TransferBits = GL_PIXEL_BUFFER_BARRIER_BIT|GL_TEXTURE_UPDATE_BARRIER_BIT|GL_BUFFER_UPDATE_BARRIER_BIT|GL_FRAMEBUFFER_BARRIER_BIT;
constexpr GLbitfield HostBits = /*GL_BUFFER_UPDATE_BARRIER_BIT|*/IOpenGL_FunctionTable::CLIENT_MAPPED_BUFFER_BARRIER_BIT;

constexpr GLbitfield AllComputeBits = GL_COMMAND_BARRIER_BIT|AnyShaderStageCommonBits;
constexpr GLbitfield AllLateFragmentTestsBits= GL_QUERY_BUFFER_BARRIER_BIT|GL_FRAMEBUFFER_BARRIER_BIT;
constexpr GLbitfield AllRasterizationBits = AnyShaderStageCommonBits|AllLateFragmentTestsBits;
constexpr GLbitfield AllProgrammablePullingBits = GL_TRANSFORM_FEEDBACK_BARRIER_BIT|AllRasterizationBits;
constexpr GLbitfield AllDirectGraphicsBits = VertexInputBits|AllProgrammablePullingBits;
constexpr GLbitfield AllGraphicsBits = GL_COMMAND_BARRIER_BIT|AllDirectGraphicsBits;
            
const GLbitfield consumerBits[PipelineStageCount] = { // dst stage mask
GL_ALL_BARRIER_BITS, // every later stage could consume anything
AllGraphicsBits|AllComputeBits, // every later shader count consume anything
AllDirectGraphicsBits, // vertex input stage is later than command stage
AllProgrammablePullingBits, // vertex shader stage is later than vertex input stage
AllProgrammablePullingBits, // control shader stage is later than vertex shader stage, but cannot consume any less resource types
AllProgrammablePullingBits, // evaluation shader stage is later than control shader stage, but cannot consume less resource types
AllProgrammablePullingBits, // geometry shader stage is later than evaluation shader stage, but cannot consume any less resource types
AllRasterizationBits, // fragment shader stage needs to include all the bits late fragment test needs
AllRasterizationBits, // early fragment test stage is later than geometry shader stage and ergo no longer writes into transform feedback can no longer occur
AllLateFragmentTestsBits, // late fragment test stage needs access to framebuffer and query buffer
GL_FRAMEBUFFER_BARRIER_BIT, // color attachment output stage only writes to framebuffer
GL_COMMAND_BARRIER_BIT|AnyShaderStageCommonBits, // compute shader stage is later than command stage
TransferBits,
0 // bottom of pipe produces nothing
};

            GLbitfield barrier = 0;
            if (flags & asset::EPSF_HOST_BIT)
                barrier |= HostBits;
            if (flags & asset::EPSF_ALL_GRAPHICS_BIT)
                barrier |= AllGraphicsBits;
            if (flags & asset::EPSF_ALL_COMMANDS_BIT)
                barrier |= AllGraphicsBits|AllComputeBits|TransferBits; // OpenGL queue can do everything
            for (uint32_t i = 0u; i < PipelineStageCount; ++i)
                if (flags & (1u << i))
                    barrier |= bits[i];
            return barrier;
```

Reply:
Then from the producer and consumer barrier bits you can figure out the actual possible bits by `srcTranslated&dstTranslated`

Reply:
for command buffer submission semaphore flags, assume src stage is bottom of pipe

Reply:
might have misunderstood Vulkan memory model https://discord.com/channels/427551838099996672/427803486269210627/871146475256676393